substrate-archive/src/wasm_tracing.rs

```diff
@@ -0,0 +1,332 @@
+		self.block_num
+	}
+}
+
+#[derive(Debug)]
+pub struct SpansAndEvents {
+	pub spans: Vec<SpanMessage>,
+	pub events: Vec<EventMessage>,
+}
+
+/// Collects traces and filters based on target.
+/// The Subscriber implementation is blocking. It uses Mutex primitives to coalesce traces before
+/// sending them to the appropriate actor.
+/// Therefore, one must be careful not to block the async executor when adding tracing spans
+/// using this subscriber implementation anywhere inside an async context in substrate-archive.
+pub struct TraceHandler {
+	/// the Block Number this trace set is from
+	block_num: u32,
+	/// Hash of the block for this trace set
+	hash: BlockHash,
+	span_events: Arc<Mutex<SpansAndEvents>>,
+	targets: Vec<(String, Level)>,
+	counter: AtomicU64,
+	current_span: CurrentSpan,
+}
+
+impl TraceHandler {
+	pub fn new(targets: &str, block_num: u32, hash: BlockHash, span_events: Arc<Mutex<SpansAndEvents>>) -> Self {
+		let targets = targets.split(',').map(|s| parse_target(s)).collect();
+		// must start indexing from 1 otherwise `tracing` panics
+		let counter = AtomicU64::new(1);
+		Self { targets, counter, span_events, block_num, hash, current_span: Default::default() }
+	}
+
+	/// Formats an event as an [`EventMessage`] and stores it in the [`SpansAndEvents`] (which is sent to the [`StorageAggregator`] after the block is executed).
+	fn gather_event(&self, event: &Event<'_>, time: DateTime<Utc>) -> Result<()> {
+		let meta = event.metadata();
+		let mut values = TraceData::default();
+		event.record(&mut values);
+		let parent_id = event.parent().cloned().or_else(|| self.current_span.id());
+
+		// check if WASM traces specify a different name/target.
+		let name = values.0.remove(WASM_NAME_KEY).map(|t| t.to_string()).unwrap_or_else(|| meta.name().to_string());
+		let target =
+			values.0.remove(WASM_TARGET_KEY).map(|t| t.to_string()).unwrap_or_else(|| meta.target().to_string());
+
+		let file = values.0.remove("file").map(Into::into);
+		let line = match values.0.remove("line").map(Into::into) {
+			Some(DataType::U64(t)) => Ok(Some(t.try_into()?)),
+			None => Ok(None),
+			_ => Err(TracingError::TypeError),
+		}?;
+
+		let event = EventMessage {
+			level: *meta.level(),
+			target,
+			name,
+			parent_id,
+			values,
+			block_num: Some(self.block_num),
+			hash: Some(self.hash.clone()),
+			time,
+			file,
+			line,
+		};
+		self.span_events.lock().events.push(event);
+		Ok(())
+	}
+
+	// we need this because we don't know the values until after tracing has been executed
+	/// Returns true if a span is part of an enabled Target. Checks WASM in addition to the spans target.
+	fn is_enabled(&self, span: &SpanMessage) -> bool {
+		let wasm_target = span.values.0.get(WASM_TARGET_KEY).map(|s| s.to_string());
+
+		self.targets.iter().filter(|t| t.0.as_str() != "wasm_tracing").any(|t| {
+			let check_target = |target: &str, lvl: Level| -> bool { target.starts_with(&t.0.as_str()) && lvl <= t.1 };
```

Comment:
This looks like an overkill, why not inline the check, or make it into a bool?

```rust
let valid_target = target.starts_with(&t.0.as_str()) && lvl <= t.1;

valid_target || wasm_target.as_ref().map(|wasm_t| check_target(wasm_t, span.level)).unwrap_or(false)
```

Reply:
Yeahp, I  made it simpler. Hopefully easier to understand now too